packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -955,7 +957,10 @@ export default function parse(source: string, state: State): TemplateParseResult
                     type: IRAttributeType.Expression,
                     value: parseTemplateExpression(el, value),
                 };
-            } else if (isBooleanAttribute) {
+            } else if (
+                isUsedAsBooleanAttribute ||
+                (isBooleanAttribute(matching.name) && isGlobalHtmlAttribute(matching.name))
```

Comment:
while relying only in the attribute name, we can only make this assumption for boolean and global attributes (`hidden`).

fyi/ currently there is an issue in the template-compiler: when we have a component with an `api` property, with the same name of a boolean attribute (ex: `capture`), we error in, for ex: `<x-child capture="true"></x-child>`. `capture` is a boolean attribute of `<input type='file'>` only.

Reply:
I noticed that on the playground, `<div foo="">div</div>` renders as `<div foo>div</div>`. This is the expected behavior (https://jsbin.com/higuperebu/edit?html,output), but I wonder if we should be categorizing that as a boolean type here, instead of letting it fall through and categorizing it as a string type.

Reply:
i don't fully understand your point. can you elaborate a bit more?

Reply:
I was just wondering whether the compiler should categorize `foo=""` as a boolean attribute here, since browsers end up rendering that as a boolean attribute in the DOM.

Reply:
I'm also curious about `foo`, and why it works like that?

Reply:
[in our case](https://gus.lightning.force.com/docs/component-library/tools/playground/kqBBKnk3a/1/edit), `foo` is always treated as an attribute. for your example in lwc it actually works like that (`elm.setAttribute('foo', '')` just sets the attribute as boolean).

The interesting case, is when you use `foo` as a boolean attribute in html: `<p foo>bla</p>` generates `<p foo="true">bla</p>`.

Reply:
> in our case, foo is always treated as an attribute. for your example in lwc it actually works like that (elm.setAttribute('foo', '') just sets the attribute as boolean).

Right. What I was saying was, since both `<div foo=""></div>` and `div.setAttribute('foo', '')` result in `foo` becoming a boolean attribute, I wonder if it's more correct to have the template-compiler categorize empty string attributes as boolean attributes at this point in the logic. It doesn't make a difference today, but if we decide to make the change proposed by @pmdartus where known boolean attributes are added only if the value is truthy, then it seems we would need to make that distinction here.

> The interesting case, is when you use foo as a boolean attribute in html: <p foo>bla</p> generates <p foo="true">bla</p>.

Wow, that's totally broken! https://playground.lwcjs.org/projects/4jjgLIB0f/1/edit